### PR TITLE
Aarch64 codegen quality: handle add-negative-imm as subtract.

### PR DESCRIPTION
**Pull request by cfallin**
We often see patterns like:

```
    mov w2, #0xffff_ffff   // uses ORR with logical immediate form
    add w0, w1, w2
```

which is just `w0 := w1 - 1`. It would be much better to recognize when
the inverse of an immediate will fit in a 12-bit immediate field if the
immediate itself does not, and flip add to subtract (and vice versa), so
we can instead generate:

```
    sub w0, w1, #1
```

We see this pattern in e.g. `bz2`, where this commit makes the following
difference (counting instructions with `perf stat`, filling in the
wasmtime cache first then running again to get just runtime):

pre:

```
        992.762250      task-clock (msec)         #    0.998 CPUs utilized
               109      context-switches          #    0.110 K/sec
                 0      cpu-migrations            #    0.000 K/sec
             5,035      page-faults               #    0.005 M/sec
     3,224,119,134      cycles                    #    3.248 GHz
     4,000,521,171      instructions              #    1.24  insn per cycle
   <not supported>      branches
        27,573,755      branch-misses

       0.995072322 seconds time elapsed
```

post:

```
        993.853850      task-clock (msec)         #    0.998 CPUs utilized
               123      context-switches          #    0.124 K/sec
                 1      cpu-migrations            #    0.001 K/sec
             5,072      page-faults               #    0.005 M/sec
     3,201,278,337      cycles                    #    3.221 GHz
     3,917,061,340      instructions              #    1.22  insn per cycle
   <not supported>      branches
        28,410,633      branch-misses

       0.996008047 seconds time elapsed
```

In other words, a 2.1% reduction in instruction count on `bz2`.

### CODE REVIEW

**Comment by cfallin** [comment]
@julian-seward1 I reworked the negation/casting a bit as per our discussion -- PTAL!

**Comment by julian-seward1** [comment]
Thanks for the simplification.  It is simpler, but tbh I am still unclear about the need for the masking.  I had in mind the following; would it work?
```
assert!(twos_complement_bits <= 64);
if let Some(imm_value) = input_to_const(ctx, input) {
    let mut sign_extended =
        ((imm_value as i64) << (64 - twos_complement_bits)) >> (64 - twos_complement_bits);
    if let Some(i) = Imm12::maybe_from_u64(sign_extended as u64) {
        return (ResultRSEImm12::Imm12(i), false);
    }
    sign_extended = sign_extended.wrapping_neg();
    if let Some(i) = Imm12::maybe_from_u64(sign_extended as u64) {
        return (ResultRSEImm12::Imm12(i), true);
    }
}
(
    ResultRSEImm12::from_rse(put_input_in_rse(ctx, input, narrow_mode)),
    false,
)

```

**Comment by cfallin** [comment]
Ah, yes, I just realized that the masking doesn't actually matter: the `Imm12`'s limit is lower than the maximum value for the masked-down width, so masking to the appropriate width first will not enable any additional constant values to be represented. (In the `i8` case where that's not true, any negative value would also fit without negation.)

**Comment by cfallin** [comment]
Thanks! Updated naming to `put_input_in_rse_imm12_maybe_negated`; hopefully clearer.

Re: the sign extension: one way to think of it is that the sign extension does not affect the correctness of the add/sub, because narrower-than-64-bit values have undefined high bits. (So even adding 127 == -128 in `i8`-world is safe, because it produces the correct lowest byte.) The high bytes only matter where `Imm12` is concerned -- it may falsely decide we can't use a 12-bit immediate because (actually ignored) high bits are set. So we need to sign extend before doing the 64-bit negation, so that we get a sign-extended positive value and can recognize small negative numbers as representable. (And why not mask the first, non-negated, case then? Because the constant with `iconst` should already be within range if positive.)
